### ISSUE TITLE
Inconsistent & unexpected NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE reported

### ISSUE THREAD

**Issue by iloveeclipse** [spotbugs/spotbugs#3573]
After merge of https://github.com/spotbugs/spotbugs/pull/3559 to master we see now **lot** of `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE` warnings because of this change for the code where the data read from the Map.get() is immediately dereferenced without NP check.

Trivial example is
```
package a;

import java.util.List;
import java.util.Map;

public class NpWarningOnMapGet {
    String key = "aa";
    public void main(Map<String, List<Object>> map) {
        if(map.get(key) != null) {
            List<Object> list = map.get(key);
            boolean b = list.contains(map); // warning here
            System.out.println(b);
        }
    }
}
```

```
Bug: Possible null pointer dereference in NpWarningOnMapGet.main(Map) due to return value of called method
 
 The return value from a method is dereferenced without a null check, and the return value of that method is one that should generally be checked for null.  This may lead to a 
NullPointerException when the code is executed.  
 
Rank: Troubling (13), 
confidence: Normal

Pattern: NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE 

Type: NP, 
Category: STYLE (Dodgy code)

Dereferenced at NpWarningOnMapGet.java:[line 11]
In method com.advantest.itee.measurement.ui.NpWarningOnMapGet.main(Map)
Value loaded from list
Known null at NpWarningOnMapGet.java:[line 10]
```

The problem with that is:  it is only reported for **some** few cases and it is not clear why. We have ~251 warnings reported on 32856 classes with 2054585 code size (as reported by SpotBugs). I'm pretty sure there is a LOT of code that reads something from the maps without checking, but we only see small fraction reported.

I've put same class above in an empty project and it seem to not report any warnings, so it seems that other classes somehow affect the analysis result...

I personally dislike the style above, and for sure the result from the map.get() has to be checked for null differently (not by executing get(key) two times, because map content may change inbetween), but why I've marked this as "bug" is 

1. Because the developers should be able to understand why the problem is reported here but not there on almost **identical** code. It should be "deterministic" to the user.
2. The code style is very common, but bug reported only for few cases. Either it should be always reported or it shouldn't be reported at all.

**Comment by iloveeclipse** [comment]
@gtoison : any hint is welcome.

**Comment by gtoison** [comment]
Would the problematic value be (on some path) the return of a method annotated with `@CheckForNull`?

**Comment by iloveeclipse** [comment]
That is really literally the code above, placed in ome existing package. I will try to find something reproducible next week. So far all the code I've analysed carried no annotations, although they are present on the classpath.

**Comment by iloveeclipse** [comment]
Could it be, the detector "sees" some code in other classes that checks the values retrieval from maps for null and assumes that all Map.get() calls should be checked too?

**Comment by gtoison** [comment]
I had a doubt about the change in the static method `IsNullValue.merge()`
Could you please try removing this bit?

        if (a.isCheckForNull() || b.isCheckForNull()) {
            combinedFlags |= CHECK_FOR_NULL_VAL;
        }

I've also seen several places where SpotBugs tries really hard to avoid allocating memory by recycling objects, it is quite possible that some of these recycled objects are not reset properly after the change I've made

**Comment by gtoison** [comment]
It's also quite possible that I misundertood how the `IsNullValue` objects are stored in `IsNullValue.instanceByFlagsList` and `mergeMatrix`

**Comment by gtoison** [comment]
I've looked into it but have no clue what might be causing the problem, it's going to be hard without a way to reproduce it.
Are the unexpected bugs all on `Map.get()` calls?

**Comment by iloveeclipse** [comment]
Currently debugging this, will take more time. 


> Are the unexpected bugs all on Map.get() calls?

Yes, at least all of samples I've checked wrre of that kind.

So far, if I let SpotBugs analyze only one class in the package, it doesn't report warnings.
As soon as whole package is analyzed (lot of other classes), the warning is reported.
More: I see that if one class checks the value of `Map.get()` for null, other class gets the warning *un fully unrelated code* on `Map.get()`. However there seem to be some dependency graph needed between classes which I can't understand yet.

**Comment by iloveeclipse** [comment]
OK, I believe I have found common root cause. All the affected projects/classes were using Guava `BiMap` or similar types.
https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/BiMap.java#L46

Once analysis saw **any** use of `BiMap` in the code, fully unrelated `Map` uses were affected.
Here is the smallest example I could produce:

```
import java.util.Map;
import com.google.common.collect.HashBiMap;

public class NpWarningOnMapGet {
	static final boolean DUMMY_ONLY_TO_USE_BIMAP = HashBiMap.create() != null? true : false;

	public boolean doSomethingCompletelyDifferent(Map<String, String> map) {
		boolean b = true;
		if(map.get("") != null) {
			b = map.get("").contains("42"); // NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE warning here
		}
		return b;
	}
}
```

Reverting 6a4ece576406cad86999e7942bcba691bd5e99d2 fixes the warning.

Note: guava `com.google.common.collect` package is annotated with `CheckReturnValue` which probably is the point here.

https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/package-info.java#L122

I honestly don't have the knowledge how Spotbugs NP analysis works in this case, but it seem the analysis now transfers the knowledge that **some** `Map` is annotated with `CheckReturnValue` to **all** (also not annotated and fully unrelated) `Map` uses. 

I'm definitely not a friend of the code style above and would never write it this way. In theory, every Map.get() dereference without NP check should be always warned. However, I believe this would cause too much of warnings now without any clear indication why, if the analysis found possibility that **some** type that extends base type is annotated with some kind of "Check..." annotation (not even directly but via package-info).

Also it is not only about `Map`, I assume any use of NP annotated code that extends some common interface together with unrelated code that simply uses base interface would now lead to undesirable warnings.

**Comment by gtoison** [comment]
Thank you very much for finding a way to reproduce, I'll look into it.
Everytime I feel like I got a better understanding of SpotBugs inner working, I immediately get the feeling that I'm barely scratching the surface!

**Comment by iloveeclipse** [comment]
I'm really undecided here because one shouldn't write such code, and if the project uses some implementation type that is annotated with NP annotations, it **may** appear at runtime in the code that uses super interface in signatures and it may run into NPE.

However we are not sure about that so probably we should only report with lowest confidence/rank/prio here. Also bug pattern description could add some hints like "... we found some NP annotated code in the project that implements XYZ interface used in this method. Make sure all uses of this interface are safe...".

So if you could reduce confidence/rank/prio for that specific case, it would probably be a solution.
Not sure if this can be done for 4.9.4 release, if not, I would vote for a temporary revert or some kimd of "mute" workaround, so that we can get 4.9.4 out of the door soon.

**Comment by gtoison** [comment]
I agree that these warnings seem to have some merit... 
I'm trying to better understand the rationale and how/where SpotBugs decides that calls are problematic 

What exact version of Guava are you using? 
They have changed the nullness annotations over time and for now I can't seem to reproduce the problem with the sample you have provided

**Comment by iloveeclipse** [comment]
> What exact version of Guava are you using?

Definitely not the latest one. Will check tomorrow.

**Comment by iloveeclipse** [comment]
I have a reproducer that doesn't need guava and also shows same problem:

```
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.CheckForNull;

public class NpWarningOnMapGet {
	static final boolean DUMMY_ONLY_TO_USE_BIMAP = DummyMap.createDummyMap() != null ? true : false;
	static final boolean DUMMY_ONLY_TO_USE_LISTS = DummyList.createDummyList().get(0) != null ? true : false;

	public boolean doSomething() {
		return doSomethingCompletelyDifferent(Map.of()) && doSomethingCompletelyDifferent(List.of());
	}

	public boolean doSomethingCompletelyDifferent(Map<String, String> map) {
		boolean b = true;
		if (map.get("") != null) {
			b = map.get("").contains("42"); // warning here
		}
		return b;
	}

	public boolean doSomethingCompletelyDifferent(List<String> list) {
		boolean b = true;
		if (list.get(0) != null) {
			b = list.get(0).contains("42"); // warning here
		}
		return b;
	}
}

class DummyList<T> extends ArrayList<T> {
	private static final long serialVersionUID = 1L;

	public static <T> DummyList<T> createDummyList() {
		return new DummyList<>();
	}

	@Override
	@CheckForNull
	public T get(int index) {
		return super.get(index);
	}
}

class DummyMap<K, V> extends HashMap<K, V> {
	private static final long serialVersionUID = 1L;

	public static <K, V> DummyMap<K, V> createDummyMap() {
		return new DummyMap<>();
	}

	@Override
	@CheckForNull
	public V get(@CheckForNull Object key) {
		return super.get(key);
	}
}
```
